File: Water Jug/waterJUG.cpp
#include <bits/stdc++.h>
#define pii pair<int, int>
#define mp make_pair
using namespace std;

void BFS(int a, int b, int target)
{
    // Map is used to store the states, every
    // state is hashed to binary value to
    // indicate either that state is visited
    // before or not
    map<pii, int> m;
    bool isSolvable = false;
    vector<pii> path;

    queue<pii> q;   // queue to maintain states
    q.push({0, 0}); // Initialing with initial state

    while (!q.empty())
    {
        bool p = false;
        pii u = q.front(); // current state

        q.pop(); // pop off used state

        // if this state is already visited
        if (m[{u.first, u.second}] == 1)
            continue;

        // doesn't met jug constraints
        if ((u.first > a || u.second > b || u.first < 0 || u.second < 0))
            continue;

        // filling the vector for constructing
        // the solution path
        path.push_back({u.first, u.second});

        // marking current state as visited
        m[{u.first, u.second}] = 1;

        // if we reach solution state, put ans=1
        if (u.first == target || u.second == target)
        {
            isSolvable = true;
            if (u.first == target)
            {
                if (u.second != 0)
                    // fill final state
                    path.push_back({u.first, 0});
            }
            else
            {
                if (u.first != 0)
                    // fill final state
                    path.push_back({0, u.second});
            }

            // print the solution path
            int sz = path.size();

            for (int i = 0; i < sz; i++)
                cout << "(" << path[i].first << ", " << path[i].second << ")\n";

            break;
        }

        // if we have not reached final state
        if (u.first < 4)
        {
            q.push({a, u.second}); // fill Jug1
            // cout << u.first << "    fill a   " << u.second << endl;
        }
        if (u.second < 3)
        {
            q.push({u.first, b}); // fill Jug2
            // cout << u.first << "    fill b  " << u.second << endl;
        }
        // empty
        if (u.second > 0)
        {
            q.push({u.first, 0}); // Empty Jug2
        }
        if (u.first > 0)
        {
            q.push({0, u.second}); // Empty Jug1
        }
        // start
        int x = u.first;
        int y = u.second;
        // pour water jug2 --> jug1
        if (((x + y) > 0) && ((x + y) >= a) && (y > 0) && (x > 0))
        {
            int aa = a;
            int bb = y - (a - x);
            q.push({aa, bb});
            // cout << aa << "+++" << bb << endl;
        }
        // pour water jug1 --> jug2
        if (((x + y) > 0) && ((x + y) >= b) && (x > 0) && (y > 0))
        {

            int bb = b;
            int aa = x - (b - y);
            q.push({aa, bb});
            cout << aa << "+-+" << bb << endl;
        }
        // // (x+y,0)
        if ((0 < (x + y) && (x + y) <= a) && y >= 0)
        {
            int aa = x + y;
            int bb = 0;
            q.push({aa, bb});
            // cout << aa << " -- " << bb << endl;
        }

        // // (0,x+y)
        if ((0 < (x + y) && (x + y) <= b) && x >= 0)
        {
            int bb = x + y;
            int aa = 0;
            q.push({aa, bb});
            // cout << aa << "  " << bb << endl;
        }
        // end
    }

    // No, solution exists if ans=0
    if (!isSolvable)
        cout << "No solution";
}

// Driver code
int main()
{
    int Jug1 = 4, Jug2 = 3, target = 2;
    cout << "Path from initial state "
            "to solution state :\n";
    BFS(Jug1, Jug2, target);
    return 0;
}

// g++ -o waterJUG waterJUG.cpp && waterJUG.exe

File: Water Jug/waterJug.py
# water jug problem
from collections import deque


def waterjug(jug1, jug2, goal):
    # dictonary  for store unique value
    map = {}
    # queue store the and dequeue --> list
    queue = deque()
    # path to store the solutation path -->list
    path = []
    # initial path (0,0)
    queue.append((0, 0))

    while queue:
        first = queue[0]  # store the first element of the queue
        # print("f", first)
        queue.popleft()
        # queue.pop(0)  # pop the element
        # check the node visited or not
        tup = tuple(first)
        if map.get(tup) == 1:
            # print(map.get(tup))
            continue

        # meet the jug
        if (first[0] > jug1) or (first[1] > jug2) or (first[0] < 0) or (first[1] < 0):
            continue
        # if not get the goal
        # 1.fill the jug2 and jug1 keep at it is
        queue.append((first[0], jug2))
        # 2.fill the first jug1 and jug2 keep at it is
        queue.append((jug1, first[1]))
        # print(f'queue ==>', queue)

        path.append(first)  # add into the main path
        map[tuple(first)] = 1  # marked as visited
        # print(f'path ==> ', path)

        # check the get the destination or not
        if (first[0] == goal) or (first[1] == goal):
            print("get the solutions", first[0], first[1])
            print(queue)
            break

        # 3.pour water jug1 to ju2 and jug2 to jut1 #todo
        for i in range(max(jug1, jug2)):

            # // pour amount ap from Jug2 to Jug1
            c = first[0] + i
            d = first[1] - i

            # // check if this state is possible or not
            if (c == jug1) or (d == 0 and d >= 0):
                queue.append((c, d))
            # // Pour amount ap from Jug 1 to Jug2
            c = first[0] - i
            d = first[1] + i

            # // check if this state is possible or not
            if (c == 0 and c >= 0) or (d == jug2):
                print(c, d)
                queue.append((c, d))

        # 4.empty jug1 (0,y)
        queue.append((0, jug2))
        # 5.empty jug2 (x,0)
        queue.append((jug1, 0))


waterjug(4, 3, 2)


# u = first
